devicetypes/stelpro/stelpro-ki-thermostat.src/stelpro-ki-thermostat.groovy

```diff
@@ -0,0 +1,572 @@
+
+	map
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.associationv2.AssociationReport cmd) {
+	delayBetween([
+		zwave.associationV1.associationRemove(groupingIdentifier:1, nodeId:0).format(),
+		zwave.associationV1.associationSet(groupingIdentifier:1, nodeId:[zwaveHubNodeId]).format(),
+		poll()
+	], 2300)
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.thermostatmodev2.ThermostatModeSupportedReport cmd) {
+	log.debug "Zwave event received: $cmd"
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.basicv1.BasicReport cmd) {
+	log.debug "Zwave event received: $cmd"
+}
+
+def zwaveEvent(physicalgraph.zwave.Command cmd) {
+	log.warn "Unexpected zwave command $cmd"
+}
+
+def refresh() {
+	poll()
+}
+
+def configure() {
+	unschedule(scheduledUpdateWeather)
+	if (settings.zipcode) {
+		runEvery1Hour(scheduledUpdateWeather)
+	}
+	poll()
+}
+
+def quickSetHeat(degrees) {
+	setHeatingSetpoint(degrees, 0)
+}
+
+def setHeatingSetpoint(preciseDegrees, delay = 0) {
+	def degrees = new BigDecimal(preciseDegrees).setScale(1, BigDecimal.ROUND_HALF_UP)
+	log.trace "setHeatingSetpoint($degrees, $delay)"
+	def deviceScale = state.scale ?: 1
+	def deviceScaleString = deviceScale == 2 ? "C" : "F"
+	def locationScale = getTemperatureScale()
+	def p = (state.precision == null) ? 1 : state.precision
+
+	def convertedDegrees
+	if (locationScale == "C" && deviceScaleString == "F") {
+		convertedDegrees = celsiusToFahrenheit(degrees)
+	}
+	else if (locationScale == "F" && deviceScaleString == "C") {
+		convertedDegrees = fahrenheitToCelsius(degrees)
+	}
+	else {
+		convertedDegrees = degrees
+	}
+
+	sendEvent(name: "heatingSetpoint", value: degrees, unit: locationScale, data: [heatingSetpointRange: heatingSetpointRange])
+	sendEvent(name: "thermostatSetpoint", value: degrees, unit: locationScale, data: [thermostatSetpointRange: thermostatSetpointRange])
+
+	delayBetween([
+		zwave.thermostatSetpointV2.thermostatSetpointSet(setpointType: 1, scale: deviceScale, precision: p, scaledValue: convertedDegrees).format(),
+		zwave.thermostatSetpointV2.thermostatSetpointGet(setpointType: 1).format(),
+		poll()
+	], 1000)
+}
+
+def quickSetOutTemp(degrees) {
+	setOutdoorTemperature(degrees, 0)
+}
+
+def setOutdoorTemperature(degrees, delay = 0) {
+	setOutdoorTemperature(degrees.toDouble(), delay)
+}
+
+def setOutdoorTemperature(Double degrees, Integer delay = 0) {
+	def deviceScale
+	def locationScale = getTemperatureScale()
+	def p = (state.precision == null) ? 1 : state.precision
+	if (locationScale == "C") {
+		deviceScale = 0
+	}
+	else {
+		deviceScale = 1
+	}
+	log.debug "setOutdoorTemperature: ${degrees}"
+	zwave.sensorMultilevelV3.sensorMultilevelReport(sensorType: 1, scale: deviceScale, precision: p, scaledSensorValue: degrees).format()
+}
+
+def increaseHeatSetpoint() {
+	float currentSetpoint = device.currentValue("heatingSetpoint")
+	def locationScale = getTemperatureScale()
+	float maxSetpoint
+	float step
+
+	if (locationScale == "C") {
+		maxSetpoint = 30;
```

Comment:
there's already a function that has these magic numbers, we should be using it instead